hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReader.java

```diff
@@ -0,0 +1,178 @@
+    this.fileInfoBlock = blockReader.nextBlock(HFileBlockType.FILE_INFO);
+
+    this.isMetadataInitialized = true;
+  }
+
+  /**
+   * Seeks to the key to look up.
+   *
+   * @param key Key to look up.
+   * @return The {@link KeyValue} instance in the block that contains the exact same key as the
+   * lookup key; or {@link null} if the lookup key does not exist.
+   * @throws IOException upon error.
+   */
+  public KeyValue seekTo(KeyValue key) throws IOException {
+    BlockIndexEntry lookUpKey = new BlockIndexEntry(key, -1, -1);
+    int rootLevelBlockIndex = searchBlockByKey(lookUpKey);
+    if (rootLevelBlockIndex < 0) {
+      // Key smaller than the start key of the first block
+      return null;
+    }
+    BlockIndexEntry blockToRead = blockIndexEntryList.get(rootLevelBlockIndex);
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, blockToRead.getOffset(), blockToRead.getOffset() + (long) blockToRead.getSize());
+    HFileDataBlock dataBlock = (HFileDataBlock) blockReader.nextBlock(HFileBlockType.DATA);
+    return seekToKeyInBlock(dataBlock, key);
+  }
+
+  /**
+   * Reads the HFile major version from the input.
+   *
+   * @param bytes  Input data.
+   * @param offset Offset to start reading.
+   * @return Major version of the file.
+   */
+  public static int readMajorVersion(byte[] bytes, int offset) {
+    int ch1 = bytes[offset] & 0xFF;
+    int ch2 = bytes[offset + 1] & 0xFF;
+    int ch3 = bytes[offset + 2] & 0xFF;
+    return ((ch1 << 16) + (ch2 << 8) + ch3);
+  }
+
+  /**
+   * Reads and parses the HFile trailer.
+   *
+   * @param stream   HFile input.
+   * @param fileSize HFile size.
+   * @return {@link HFileTrailer} instance.
+   * @throws IOException upon error.
+   */
+  private static HFileTrailer readTrailer(FSDataInputStream stream,
+                                          long fileSize) throws IOException {
+    int bufferSize = HFileTrailer.getTrailerSize();
+    long seekPos = fileSize - bufferSize;
+    if (seekPos < 0) {
+      // It is hard to imagine such a small HFile.
+      seekPos = 0;
+      bufferSize = (int) fileSize;
+    }
+    stream.seek(seekPos);
+
+    byte[] byteBuff = new byte[bufferSize];
+    stream.readFully(byteBuff);
+
+    int majorVersion = readMajorVersion(byteBuff, bufferSize - 3);
+    int minorVersion = byteBuff[bufferSize - 3];
+
+    HFileTrailer trailer = new HFileTrailer(majorVersion, minorVersion);
+    trailer.deserialize(new DataInputStream(new ByteArrayInputStream(byteBuff)));
+    return trailer;
+  }
+
+  /**
+   * Searches the block that may contain the lookup key based the starting keys
+   * of all blocks (sorted in the input list), using binary search.
+   *
+   * @param lookUpKey The key to lookup.
+   * @return Block index in the input. An index outside the range of input means the key does not
+   * exist in the HFile.
+   */
+  private int searchBlockByKey(BlockIndexEntry lookUpKey) {
```

Comment:
Is this score mostly reused, or written from scratch?

Reply:
This written from scratch.